File: code/java/lists/src/main/java/org/swufe/datastructure/CircularLinkedList.java
package org.swufe.datastructure;

import java.util.Iterator;
import java.util.NoSuchElementException;

/**
 * A circularly linked list.
 */
public class CircularLinkedList<Item> implements Iterable<Item> {
    private static class Node<Item> {
        Item item;
        Node<Item> next;
        Node(Item item) {
            this(item, null);
        }

        Node(Item item, Node<Item> next) {
            this.item = item;
            this.next = next;
        }
    }

    private Node<Item> tail; // we store tail (but not head)
    private int size;

    public CircularLinkedList() {
        tail = null;
        size = 0;
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public Item first() {
        if (size == 0) {
            return null;
        }
        return tail.next.item;
    }

    public Item last() {
        if (size == 0) {
            return null;
        }
        return tail.item;
    }

    public void addFirst(Item item) {
        if (size == 0) {
            tail = new Node<>(item);
            tail.next = tail; // link to itself circularly
        } else {
            tail.next = new Node<>(item, tail.next);
        }
        size += 1;
    }

    public void addLast(Item item) {
        addFirst(item);
        tail = tail.next;
    }

    public void removeFirst() {
        if (size == 0) {
            throw new NoSuchElementException();
        }
        Node<Item> head = tail.next;
        if (head == tail) {
            // only one element
            tail = null;
        } else {
            tail.next = head.next;
        }
        size -= 1;
    }

    public void removeLast() {
        if (size == 0) {
            throw new NoSuchElementException();
        }
        Node<Item> head = tail.next;
        if (head == tail) {
            // only one element
            tail = null;
        } else {
            // find the second node to last
            while (head.next != tail) {
                head = head.next;
            }
            head.next = tail.next;
            tail = head;
        }
        size -= 1;
    }

    // rotate the first element to the back of the list
    public void rotate() {
        if (tail != null) {
            tail = tail.next;
        }
    }

    @Override
    public Iterator<Item> iterator() {
        return new LinkedIterator();
    }

    private class LinkedIterator implements Iterator<Item> {
        Node<Item> node;
        int i;

        LinkedIterator() {
            this.i = 0;
            if (size > 0) {
                node = tail.next;
            } else {
                node = null;
            }
        }

        @Override
        public boolean hasNext() {
            return i < size;
        }

        @Override
        public Item next() {
            Item answer = node.item;
            node = node.next;
            i++;
            return answer;
        }
    }

}

File: code/python/lists/circular_linked_list.py
from no_element import NoElement


class CircularLinkedList:
    """A circularly linked list."""

    class Node:
        def __init__(self, item, nxt=None):
            self.item = item
            self.next = nxt

    def __init__(self):
        self._tail = None
        self._size = 0

    def size(self):
        return self._size

    def is_empty(self):
        return self._size == 0

    def first(self):
        if self.is_empty():
            return None
        return self._tail.next.item

    def last(self):
        if self.is_empty():
            return None
        return self._tail.item

    def add_first(self, item):
        if self.is_empty():
            self._tail = self.Node(item)
            self._tail.next = self._tail
        else:
            self._tail.next = self.Node(item, self._tail.next)
        self._size += 1

    def add_last(self, item):
        self.add_first(item)
        self._tail = self._tail.next

    def remove_first(self):
        if self.is_empty():
            raise NoElement
        head = self._tail.next
        if head is self._tail:
            # only one element
            self._tail = None
        else:
            self._tail.next = head.next
        self._size -= 1

    def remove_last(self):
        if self.is_empty():
            raise NoElement
        head = self._tail.next
        if head is self._tail:
            # only one element
            self._tail = None
        else:
            # find the second to the last
            while head.next is not self._tail:
                head = head.next
            head.next = self._tail.next
            self._tail = head
        self._size -= 1

    def rotate(self):
        if not self.is_empty():
            self._tail = self._tail.next

    def __iter__(self):
        return self.LinkedIterator(self._tail, self._size)

    class LinkedIterator:
        def __init__(self, tail, size):
            self.i = 0
            self.size = size
            if size > 0:
                self.node = tail.next
            else:
                self.node = None

        def __next__(self):
            if self.i >= self.size:
                raise StopIteration
            answer = self.node.item
            self.node = self.node.next
            self.i += 1
            return answer
